File: 0084) Largest Rectangle in Histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int n = heights.size();
        stack<pair<int, int>> st;
        int ans = -1;
        int id = 0;
        for(int i = 0; i < n; i++)
        {
            id = i;
            while(!st.empty() && st.top().second > heights[i])
            {
                ans = max(ans, (i - st.top().first) * st.top().second);
                id = st.top().first;
                st.pop();
            }
            st.push({id, heights[i]});
        }
        while(!st.empty())
        {
            ans = max(ans, (n - st.top().first) * st.top().second);
            st.pop();
        }
        return ans;
    }
};

File: 0084) Largest Rectangle in Histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        int n = heights.length;
        Stack<Pair> st = new Stack<>();
        int id = 0;
        int ans = -1;
        for(int i = 0; i < n; i++)
        {
            id = i;
            while(!st.empty() && st.peek().y > heights[i])
            {
                Pair p = st.pop();
                ans = Math.max(ans, (i - p.x) * p.y);
                id = p.x;
            }
            st.push(new Pair(id, heights[i]));
        }
        while(!st.empty())
        {
            Pair p = st.pop();
            ans = Math.max(ans, (n-p.x) * p.y);
        }
        return ans;
    }
}

class Pair{
    int x;
    int y;
    Pair(int x, int y)
    {
        this.x = x;
        this.y = y;
    }
}
